### PR TITLE
Pushed some example changes, adding comments in the file changes

### PR DESCRIPTION
**Pull request by chandlerprall**

### CODE REVIEW

File: src/components/button/button_toggle/button_toggle.test.tsx [hunk 1]
```diff
@@ -8,6 +8,7 @@ describe('EuiButtonToggle', () => {
   test('is rendered', () => {
     const component = render(
       // @ts-ignore @chandler
+      // `onClick` is marked as required in EuiButtonTogglePropsForButtonToggle
       <EuiButtonToggle {...requiredProps} label="Label me" />
```
Comment (chandlerprall) [inline src/components/button/button_toggle/button_toggle.test.tsx:12]: `onClick` is required as defined by

```
type EuiButtonTogglePropsForButtonToggle = EuiButtonToggleProps &
  ButtonHTMLAttributes<HTMLButtonElement> & {
    onClick: MouseEventHandler<HTMLButtonElement>;
    value?: string;
  };
```

This same configuration affects `EuiButtonIcon` and its tests

```
type EuiButtonIconPropsForButton = EuiButtonIconProps &
  ButtonHTMLAttributes<HTMLButtonElement> & {
    onClick: MouseEventHandler<HTMLButtonElement>;
    buttonRef?: Ref<HTMLButtonElement>;
  };
```

File: src/components/button/button_toggle/button_toggle.tsx [hunk 2]
```diff
@@ -113,7 +112,7 @@ export const EuiButtonToggle: FunctionComponent<Props> = ({
         color={color}
         disabled={isDisabled}
         size={isIconOnly ? 's' : undefined} // only force small if it's the icon only version
-        {...rest}>
+        {...rest as Extract<EuiButtonTogglePropsForAnchor, EuiButtonTogglePropsForButtonToggle>}>
```
Comment (chandlerprall) [inline src/components/button/button_toggle/button_toggle.tsx:117]: TS's issue stems from it not understanding that `EuiButtonToggle`'s discriminator is the same as `EuiButton`'s, so it tries to match all the props button toggle _could_ have (e.g. `autoFocus`) to all variations in button.

`Extract<EuiButtonTogglePropsForAnchor, EuiButtonTogglePropsForButtonToggle>` returns only props that exist on both sides, satisfying `EuiButton`'s props.

TS is okay with this casting trick due to its usage of bivariance.

File: src/components/button/button_toggle/button_toggle.tsx [hunk 3]
```diff
@@ -41,16 +41,18 @@ export interface EuiButtonToggleProps extends EuiButtonProps {
    * Used primarily for `EuiButtonGroup`
    */
   type?: ToggleType;
+
+  onChange?: React.FormEventHandler<HTMLInputElement>;
```
Comment (chandlerprall) [inline src/components/button/button_toggle/button_toggle.tsx:45]: `onChange` below is being applied to the `input` element but the `onChange` prop was inheriting from both `HTMLAnchorElement` and `HTMLButtonElement`.

File: src/components/button/button.tsx [hunk 4]
```diff
@@ -160,8 +161,8 @@ export const EuiButton: FunctionComponent<Props> = ({
         href={href}
         target={target}
         rel={secureRel}
-        ref={buttonRef}
-        {...rest}>
+        ref={buttonRef as Ref<HTMLAnchorElement>}
```
Comment (chandlerprall) [inline src/components/button/button.tsx:164]: TS isn't intelligent enough to follow the impact of the union's discrimination through the rest of the function. We do the same hack at https://github.com/elastic/eui/blob/651c8b399f7a57128b5bf8050305bc2ae9c2d69d/src/components/progress/progress.tsx#L88

@thompsongl found an issue in the TS repo related to this, I was unable to find the link in my chat history with him.

File: src/components/button/button_icon/button_icon.tsx [hunk 5]
```diff
@@ -115,9 +115,8 @@ export const EuiButtonIcon: FunctionComponent<Props> = ({
         href={href}
         target={target}
         rel={secureRel}
-        // @ts-ignore don't know what's going on here @chandler
-        ref={buttonRef}
-        {...rest}>
+        ref={buttonRef as Ref<HTMLAnchorElement>}
```
Comment (chandlerprall) [inline src/components/button/button_icon/button_icon.tsx:118]: Same hack as `button.tsx`

**Comment by pugnascotia** [comment]
Thanks for the help, I've incorporated this into my branch copy.

**Pull request closed without merging**